planner/core/logical_plan_builder.go

```diff
@@ -4244,6 +4245,55 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
 	return result, nil
 }
 
+func (ds *DataSource) extractFD() *fd.FDSet {
+	// FD in datasource (leaf node) can be cached and reused.
+	if ds.fdSet == nil {
+		fds := &fd.FDSet{}
+		allCols := fd.NewFastIntSet()
+		// should use the column's unique ID avoiding fdSet conflict.
+		for _, col := range ds.TblCols {
+			// todo: change it to int64
+			allCols.Insert(int(col.UniqueID))
+		}
+		for _, idx := range ds.tableInfo.Indices {
+			keyCols := fd.NewFastIntSet()
+
+			// consider invisible index.
+			if !idx.Invisible {
```

Comment:
We can consider this. The `invisible` only means that we cannot use it. But we can use the property to maintain the logical property.